### ISSUE TITLE
Possible memory leak in Android

### ISSUE THREAD

**Issue by vsheyanov** [react-native-maps/react-native-maps#5178]
### Summary

In our production app we're constantly receiving app crash reports in Crashlytics with "out of memory" and traces of `maps`.
I tried to reproduce the problem in a fairly simple app, but not too trivial.
https://github.com/vsheyanov/react-native-maps-memory

Real use case: you have someone who needs to visit multiple addresses a day, so you have a list of visits and when you click on a specific visit, you're redirected to a Screen with a map.

Video of a problem recorded on Android Emulator with Telemetry: 
https://drive.google.com/file/d/1pYOPre3yjZ8LpQCQWqw_B59QbP2JTOpV/view?usp=sharing
It's fairly long, so feel free to scroll through it, just check the memory consumed in the beginning if the video and in the end.
00:00 - 02:40 - clicking on list items, opening maps
02:40 - 04:20 - waiting if consumed memory goes down
Memory consumption is doubled withing 2.5 minutes.

There is a built in case just to make sure that GC does it's job - you can render ~1000 random images instead of a map. 
https://drive.google.com/file/d/18K2fIKLMIny18dL0bT3rONrL176eJaIb/view?usp=sharing
In this case you can see how memory consumption goes UP, but it always goes down to a level where it was before navigating to a Screen.

So the question is - is it a memory leak in the library that causes GC not to work or is it something else?

### Reproducible sample code

```TSX
import './gesture-handler';

import React, {  useEffect } from 'react';
import { Button, FlatList, Image, SafeAreaView, StyleSheet, Text } from 'react-native';
import { createStackNavigator } from '@react-navigation/stack';
import { NavigationContainer } from '@react-navigation/native';
import MapView, { PROVIDER_GOOGLE } from 'react-native-maps';
import { ScrollView } from 'react-native-gesture-handler';

function App({ navigation }): React.JSX.Element {
  return (
    <SafeAreaView>
      <FlatList
        data={new Array(100).fill(0).map((_, index) => index)}
        renderItem={(value) => (
          <Button title={`Item ${value.item}`} onPress={() => navigation.navigate('MemoryLeak', { value: value.item })} />
        )}/>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  map: {
    ...StyleSheet.absoluteFillObject,
    marginTop: 60,
  },
});

const Map = ({ route, navigation }) => {
  useEffect(() => {
    console.log('Map mount');
    return () => {
      console.log('Map unmount');
    };
  }, []);

  return (
    <>
      <Button title="Reset stack" onPress={() => navigation.reset({ routes: [{ name: 'Home'}]})} />
      <Text>{route.params.value}</Text>
      <MapView
        provider={PROVIDER_GOOGLE}
        style={styles.map}
        region={{
          latitude: 37.78825,
          longitude: -122.4324,
          latitudeDelta: 0.015,
          longitudeDelta: 0.0121,
        }}/>
    </>
  );
};

const ImageList = () => (
  <ScrollView>
    {
      new Array(1000).fill(0).map((_, index) => (
        <Image
          key={index}
          source={{ uri: `https://picsum.photos/200/300?random=${Math.round(Math.random() * 1000)}` }}
          style={{ width: 300, height: 400}} />
      ))
    }
  </ScrollView>
);


const Stack = createStackNavigator();

const renderMap = true;

const MainApp = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator initialRouteName="Home">
        <Stack.Screen name="Home" component={App} />
        <Stack.Screen name="MemoryLeak" component={renderMap ? Map : ImageList} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MainApp;
```

### Steps to reproduce

- checkout sample repository (https://github.com/vsheyanov/react-native-maps-memory)
- add Map API Key
- run the app
- run telemetry
- start clicking list items 
- observe how memory consumption grows

### Expected result

Consumed memory goes back when GC does it's job.

### Actual result

Consumed memory stays on the same level

### React Native Maps Version

1.18.0

### What platforms are you seeing the problem on?

Android

### React Native Version

0.75.4

### What version of Expo are you using?

Not using Expo

### Device(s)

Android Emulator Pixel 7, API 33

**Comment by spyshower** [comment]
While I don't have any actual data to back this up, I have definitely suspected this for at least a year.

**Comment by SohaibKtb** [comment]
+1

**Comment by zelardiel** [comment]
> While I don't have any actual data to back this up, I have definitely suspected this for at least a year.

I think the memory crashes that we have been seeing is also around that time span (maybe slightly more)

**Comment by emaxedon** [comment]
I am experiencing the same issue.

**Comment by emaxedon** [comment]
I've included a bug report of the memory leak for Android specifically (we are not experiencing it for iOS). We believe the issue is related to custom map markers for Android.

```
java.lang.OutOfMemoryError: Failed to allocate a 200 byte allocation with 1052984 free bytes and 1028KB until OOM, target footprint 268435456, growth limit 268435456; giving up on allocation because <1% of heap free after GC.
        at m.eia.g(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:79)
        at m.eia.<init>(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:43)
        at m.eib.a(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:32)
        at m.eic.j(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:60)
        at m.dmq.<init>(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:7)
        at m.dsa.a(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:19)
        at com.google.maps.api.android.lib6.phoenix.cp.B(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:174)
        at com.google.maps.api.android.lib6.phoenix.cp.n(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:6)
        at com.google.maps.api.android.lib6.phoenix.cp.h(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:198)
        at com.google.maps.api.android.lib6.impl.es.c(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:24)
        at com.google.maps.api.android.lib6.impl.ej.Z(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:8)
        at com.google.maps.api.android.lib6.impl.ej.z(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:98)
        at com.google.android.gms.maps.model.internal.u.bt(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:356)
        at m.axu.onTransact(:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:21)
        at android.os.Binder.transact(Binder.java:1219)
        at com.google.android.gms.internal.maps.zza.zzc(zza:2)
        at com.google.android.gms.internal.maps.zzab.zzt(zzab:3)
        at com.google.android.gms.maps.model.Marker.setIcon(Marker:2)
        at com.rnmaps.maps.MapMarker.updateMarkerIcon(MapMarker:295)
        at com.rnmaps.maps.MapMarker.updateCustomForTracking(MapMarker:287)
        at com.rnmaps.maps.ViewChangesTracker.update(ViewChangesTracker:64)
        at com.rnmaps.maps.ViewChangesTracker$1.run(ViewChangesTracker:22)
        at android.os.Handler.handleCallback(Handler.java:958)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:230)
        at android.os.Looper.loop(Looper.java:319)
        at android.app.ActivityThread.main(ActivityThread.java:8919)
        at java.lang.reflect.Method.invoke(Method.java:-2)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:578)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
```

**Comment by vsheyanov** [comment]
@emaxedon we do have the same thing in our app - `markers` make the memory problem worse. In my reproduction steps and video I did not add markers just to isolate `map only` case.
This a memory behavior of the app when a map is rendered with a marker(s) and I'm not even touching the app:
![image](https://github.com/user-attachments/assets/b7a9daa1-b1c4-4cf0-8e2a-efd4e64e7a73)

For markers I'll submit a separate issue with reproduction steps later (if no one does it yet).

**Comment by emaxedon** [comment]
@vsheyanov I will try `liteMode={true}` to see if this can help with the memory leak issue. I'll report back to confirm whether `liteMode={true}` can help with the memory leak. I've also added `tracksViewChanges={false}` to my markers, hoping this will help with the memory leak as well.

**Comment by emaxedon** [comment]
@vsheyanov I've seen in other threads that using the legacy renderer could solve the issue as well. I believe the memory leak currently exists in the new renderer. Have you tried with the legacy renderer setting?

I'm quite shocked frankly that this issue is still present in the library...

**Comment by robert-nash** [comment]
I don't know if this is helpful but I seem to be having exactly the same problem but the errors are happening when the app is not in the foreground (i.e. some period of time after the map has been interacted with/rendered). I haven't observed it enough to say much more than that.

**Comment by stanislav-sidorov-onlocation** [comment]
same issue
Galaxy S21 FE 5G android 14
custom markers with images
dynamic and clickable

`Exception java.lang.OutOfMemoryError:
  at m.eia.g (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:79)
  at m.eia.<init> (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:43)
  at m.eib.a (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:32)
  at m.eic.j (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:60)
  at m.dmq.<init> (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:7)
  at m.dsa.a (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:19)
  at com.google.maps.api.android.lib6.phoenix.cp.B (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:174)
  at com.google.maps.api.android.lib6.phoenix.cp.n (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:6)
  at com.google.maps.api.android.lib6.phoenix.cp.h (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:198)
  at com.google.maps.api.android.lib6.impl.es.c (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:24)
  at com.google.maps.api.android.lib6.impl.ej.Z (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:8)
  at com.google.maps.api.android.lib6.impl.ej.z (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:98)
  at com.google.android.gms.maps.model.internal.u.bt (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:356)
  at m.axu.onTransact (:com.google.android.gms.policy_maps_core_dynamite@243625205@243625202025.679729749.679729749:21)
  at android.os.Binder.transact (Binder.java:1219)
  at com.google.android.gms.internal.maps.zza.zzc (com.google.android.gms:play-services-maps@@18.2.0:2)
  at com.google.android.gms.internal.maps.zzab.zzt (com.google.android.gms:play-services-maps@@18.2.0:3)
  at com.google.android.gms.maps.model.Marker.setIcon (com.google.android.gms:play-services-maps@@18.2.0:2)
  at com.rnmaps.maps.MapMarker.updateMarkerIcon (MapMarker.java:295)
  at com.rnmaps.maps.MapMarker.updateCustomForTracking (MapMarker.java:287)
  at com.rnmaps.maps.ViewChangesTracker.update (ViewChangesTracker.java:64)
  at com.rnmaps.maps.ViewChangesTracker$1.run (ViewChangesTracker.java:22)
  at android.os.Handler.handleCallback (Handler.java:958)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loopOnce (Looper.java:230)`

**Comment by roachadam** [comment]
RN: 0.74.5
Expo: 51.0.39
react-native-maps: 1.18.2

`Exception java.lang.OutOfMemoryError:
  at java.util.Arrays.copyOf (Arrays.java:3560)
  at java.util.Arrays.copyOf (Arrays.java:3529)
  at java.util.ArrayList.toArray (ArrayList.java:376)
  at m.fyn.o (:com.google.android.gms.policy_maps_core_dynamite@244125205@244125202025.684868338.684868338:26)
  at m.eaj.f (:com.google.android.gms.policy_maps_core_dynamite@244125205@244125202025.684868338.684868338:8)
  at m.ent.n (:com.google.android.gms.policy_maps_core_dynamite@244125205@244125202025.684868338.684868338:867)
  at m.exp.run (:com.google.android.gms.policy_maps_core_dynamite@244125205@244125202025.684868338.684868338:802)`

Also seems to be happening related to custom markers or callouts. Issue is not present on IOS

**Comment by boxedcode** [comment]
I'm also experiencing the same issue, using the default marker.

RN: 0.76.3
Expo: 52.0.7
react-native-maps: 1.20.1

```
 ERROR  Your app just crashed. See the error below.
java.lang.OutOfMemoryError: Failed to allocate a 4194316 byte allocation with 3668304 free bytes and 3MB until OOM
  dalvik.system.VMRuntime.newNonMovableArray(Native Method)
  android.graphics.Bitmap.nativeCreate(Native Method)
  android.graphics.Bitmap.createBitmap(Bitmap.java:879)
  android.graphics.Bitmap.createBitmap(Bitmap.java:856)
  android.graphics.Bitmap.createBitmap(Bitmap.java:823)
  com.google.maps.api.android.lib6.gmm6.vector.gl.k.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):6)
  com.google.maps.api.android.lib6.gmm6.vector.gl.ai.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):3)
  com.google.maps.api.android.lib6.gmm6.vector.gl.ai.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):17)
  com.google.maps.api.android.lib6.gmm6.vector.gl.ai.b(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):2)
  com.google.maps.api.android.lib6.gmm6.vector.gl.drawable.af.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):10)
  com.google.maps.api.android.lib6.gmm6.vector.gl.drawable.ao.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):86)
  com.google.maps.api.android.lib6.gmm6.vector.ca.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):82)
  com.google.maps.api.android.lib6.gmm6.vector.bt.a(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):179)
  com.google.maps.api.android.lib6.gmm6.vector.av.run(:com.google.android.gms.dynamite_mapsdynamite@202414048@20.24.14 (040400-0):59)
```

**Comment by spyshower** [comment]
> [@emaxedon](https://github.com/emaxedon) we're using `lightMode={true}` for some cases (and it seems to work OK), but we do have a case when this is not possible - when we need to have a fully interactive map with a Marker that is clickable. And chart from above is exactly that case.

I think you mean `liteMode`

**Comment by ad-pereira** [comment]
Hi, is there any news about this?

**Comment by robert-nash** [comment]
Still relevant

**Comment by yunusemre** [comment]
Still relevant  👍

**Comment by piconepaolo** [comment]
Still relevant

**Comment by federicogomez-stowlog** [comment]
Still relevant

**Comment by maxedon** [comment]
Still relevant

**Comment by pdrohm** [comment]
still relevant

**Comment by marcusviniciusf** [comment]
still relevant

**Comment by HamayelAndres** [comment]
Still relevant

**Comment by TactfulEight784** [comment]
Has anyone found a solution to this issue yet or will this ever get looked at. 
This issue basically makes this library unusable on android in stacks.
I’m currently using a native stack trying to implement maps into a set of screens that are constantly navigated to and from and with this issue it causes my app to crash with an out of memory error pretty quickly. 
This is using the most recent version of the package 1.26.20

**Comment by jacksonknightwood** [comment]
We are experiencing memory leak in production iOS app. Need some help urgently.
we are using the most recent version 1.26.20

**Comment by thisisadarshjaiswal** [comment]
still relevant

**Comment by thisisadarshjaiswal** [comment]
Any solution?

**Comment by ottor-o** [comment]
Im not sure if this is helpful or not but I managed to fix a similar (or the same) issue on another library that uses google maps (react-native-google-maps-plus), ([pr #98](https://github.com/pinpong/react-native-google-maps-plus/pull/98)). On it the issue was that the MapView class instance was referenced by another class. As long as the second class wasn't nullified on destroy, reference to MapView was kept preventing it from getting garbage collected.

The memory leak was pretty noticeable on my app since I have modal with a flatlist with multiple mapviews. Even with lite mode enabled, scrolling back and forth crashed the app pretty quickly.

I used to use react-native-maps previously and experienced the exact same issue. My hunch is that something is preventing MapView instance from getting garbage collected. It's not nullified, its referenced in one or multiple classes, or there's something wrong with its lifecycle handling.

**Comment by Dylan0115** [comment]
Still Relevant on the latest version 1.27.1

**Comment by spyshower** [comment]
This library is abandoned .

**Comment by emaxedon** [comment]
Still relevant, I've set `trackViewChanges` to `false` for custom markers and I use `liteMode` to `true` to circumvent the memory leak, but our users would love to be able to pan the map on android...

**Comment by RyanxWolton** [comment]
+1, still an issue

**Comment by lucasmendesdeliver** [comment]
I was able to resolve the memory leak by applying this patch from this issue https://github.com/react-native-maps/react-native-maps/issues/5907

**Comment by llamington** [comment]
I'm noticing this exact problem

**Comment by salah-ghanim** [comment]
I will look into the example to see if it's visible, I assumed this issue would have been solved with the migration to Fabric.

**Comment by lxr-mwhite** [comment]
@salah-ghanim It would be great to get your thoughts on issue #5907 which includes a patch that has shown to help with the memory leak and also address an issue related to the map and its markers re-rendering when navigating away and coming back. 

Seems like it has addressed some issues for a few people, including my own app.

**Comment by Gogn** [comment]
We hit this in production and tracked it down to a specific lifecycle bug with Android heap dumps. Posting the analysis in case it's useful.

### Root cause

`com.google.android.gms.maps.MapView` requires `onCreate` to be called **exactly once per instance** — only `onStart`/`onResume`/`onPause`/`onStop` may repeat. `MapView.java` drives that lifecycle from the View's attach/detach callbacks, and in `onAttachedToWindow`:

```java
attachLifecycleObserver();
if (savedMapState != null) {
    super.onCreate(savedMapState);   // <-- runs again on every re-attach
    super.onStart();
    super.onResume();
```

`savedMapState` is only non-null *after* a previous detach, so this path is taken on every **re-attach** — and `react-native-screens` detaches and re-attaches the view on every navigation. Each navigation therefore constructs a brand-new `GoogleMap` renderer with its own tile cache.

The previous renderer is never torn down, because `onDetachedFromWindow` only saves instance state and calls `onPause()` / `onStop()`. `onDestroy()` is reachable **solely** from `doDestroy()`, which runs on React unmount. In our app the React component never unmounted, so `onDestroy` never ran even once — every navigation stacked another live renderer on top of one that had merely been paused.

### Heap evidence

Android hprof, app heap only, after ~20 navigations:

- **15** live `com.google.android.gms.maps.GoogleMap` instances against **2** `com.rnmaps.maps.MapView` components
- 15 copies of the internal map listener classes
- the single largest holder in the entire heap was one obfuscated Maps class at **18,791 arrays / 45.6 MB** of tile data
- Maps accounted for **89.1 MB (78.6%)** of all primitive-array bytes
- total Java heap 147 MB against the device's 192 MB `dalvik.vm.heapgrowthlimit`, ending in `OutOfMemoryError`

After guarding the `onCreate` call so it can't run twice for the same map, on an otherwise identical build and repro:

- that 45.6 MB holder **disappeared from the dump entirely**
- aggregate `int[]` fell from **70.1 MB to 29.2 MB**
- Maps' share of primitive-array bytes dropped to **41.5 MB (59.7%)**

Reproduced on **1.27.2**; the same code is present in 1.29.0. Environment: RN 0.83.6 and 0.86.3, new architecture (Fabric), `react-native-screens` 4.23–4.26.

### Relationship to #5907 and #5778

I believe both are the same root cause.

[#5907](https://github.com/react-native-maps/react-native-maps/issues/5907) states that `onDetachedFromWindow` destroys the map and proposes an `isTemporaryDetachment` guard to prevent that destruction. Per the source above, **detach never destroys the map** — it only pauses and stops it, leaving the renderer fully alive. The defect is the unconditional `super.onCreate()` on the *attach* side, so guarding destruction wouldn't stop the leak (though it would likely mask the flicker).

[#5778](https://github.com/react-native-maps/react-native-maps/issues/5778) (markers reverting to the default red pin after navigating back) also looks like this bug: markers get re-added to a fresh renderer that never received the custom icons.

### Suggested shape of a fix

Pair the lifecycle correctly: `onCreate`/`onDestroy` with React mount/unmount, and only `onStart`/`onResume`/`onPause`/`onStop` with window attach/detach. That matches the documented contract for the underlying `MapView`.

Concretely, skip `super.onCreate()` in `onAttachedToWindow` when a map already exists. One caveat worth noting for whoever implements it: `map != null` is an imperfect sentinel, because `map` isn't assigned until `onMapReady` fires — a detach/re-attach *before* the map becomes ready would slip through the guard. An explicit boolean set at the point `onCreate` is actually called is more robust.

### Patch

This is the exact change the measurements above came from, against `1.27.2`:

```diff
--- a/android/src/main/java/com/rnmaps/maps/MapView.java
+++ b/android/src/main/java/com/rnmaps/maps/MapView.java
@@ -326,9 +326,16 @@
         }
 
         attachLifecycleObserver();
         if (savedMapState != null) {
-            super.onCreate(savedMapState);
+            // The underlying Google MapView expects onCreate exactly once per
+            // instance; only onStart/onResume may repeat. This branch runs on every
+            // re-attach (savedMapState is non-null only after a detach), and calling
+            // onCreate again builds a whole new map renderer while the previous one
+            // is merely paused - onDestroy runs solely from doDestroy() on unmount.
+            if (map == null) {
+                super.onCreate(savedMapState);
+            }
             super.onStart();
             super.onResume();
             prepareAttacherView();
             getMapAsync((map)->{
```

For an actual fix I'd suggest the explicit-sentinel form instead, since `map` isn't assigned until `onMapReady` fires and a detach/re-attach before then would slip past a `map != null` check. It also covers the `onCreate(LifecycleOwner)` path:

```diff
--- a/android/src/main/java/com/rnmaps/maps/MapView.java
+++ b/android/src/main/java/com/rnmaps/maps/MapView.java
@@ -93,8 +93,17 @@
     private Bundle savedMapState;
     private ArrayList<MapFeature> savedFeatures = null;
     private boolean shouldRestorePadding = false;
 
+    /**
+     * Whether super.onCreate() has already run for this instance. The underlying
+     * Google MapView requires onCreate exactly once; onStart/onResume may repeat.
+     * Tracked explicitly rather than inferred from `map != null`, because `map` is
+     * not assigned until onMapReady fires - a detach/re-attach before the map
+     * becomes ready would otherwise create a second renderer.
+     */
+    private boolean mapCreated = false;
+
     private MarkerManager markerManager;
     private MarkerManager.Collection markerCollection;
     private PolylineManager polylineManager;
     private PolylineManager.Collection polylineCollection;
@@ -203,9 +212,12 @@
 
 
     @Override
     public void onCreate(LifecycleOwner owner) {
-        super.onCreate(null);
+        if (!mapCreated) {
+            mapCreated = true;
+            super.onCreate(null);
+        }
     }
 
     @Override
     public void onStart(LifecycleOwner owner) {
@@ -326,9 +338,12 @@
         }
 
         attachLifecycleObserver();
         if (savedMapState != null) {
-            super.onCreate(savedMapState);
+            if (!mapCreated) {
+                mapCreated = true;
+                super.onCreate(savedMapState);
+            }
             super.onStart();
             super.onResume();
             prepareAttacherView();
             getMapAsync((map)->{
```

Only Variant A has been tested; Variant B is the same idea with a tighter sentinel and hasn't been benchmarked.
